shell/platform/android/io/flutter/embedding/android/FlutterImageView.java

```diff
@@ -242,52 +245,74 @@ public void closeImageReader() {
     imageReader.close();
   }
 
+  private void setCurrentBitmap(Bitmap bitmap) {
+    synchronized (bitmapLock) {
+      currentBitmap = bitmap;
+      isBitmapAvailableForRendering = bitmap != null;
+    }
+  }
+
   @Override
   protected void onDraw(Canvas canvas) {
     super.onDraw(canvas);
-    if (currentImage != null) {
-      updateCurrentBitmap();
-    }
-    if (currentBitmap != null) {
-      canvas.drawBitmap(currentBitmap, 0, 0, null);
+    synchronized (bitmapLock) {
+      if (currentBitmap != null) {
+        canvas.drawBitmap(currentBitmap, 0, 0, null);
+        isBitmapAvailableForRendering = false;
+      }
     }
   }
 
-  private void closeCurrentImage() {
-    // Close and clear the current image if any.
-    if (currentImage != null) {
-      currentImage.close();
-      currentImage = null;
+  // This method is run on the |sBackgroundHandler|.
+  @Override
+  public void onImageAvailable(ImageReader reader) {
+    // 1. `acquireLatestImage` will discard any images in the queue up to the most recent
+    // one.
+    // 2. `acquireLatestImage()` may return null if no new image is available.
+    // 3. There's no guarantee that `onDraw()` is called after `invalidate()`.
+    // For example, the device may not produce new frames if it's in sleep mode
+    // or some special Android devices so the calls to `invalidate()` queued up
+    // until the device produces a new frame.
+    // 4. While the engine will also stop producing frames, there is a race condition.
+    try (final Image image = reader.acquireLatestImage()) {
+      if (image == null) {
+        return;
+      }
+
+      Bitmap bitmap = convertImageToBitmap(image);
+      if (bitmap != null) {
+        setCurrentBitmap(bitmap);
+        sUiHandler.postAtFrontOfQueue(() -> invalidate());
+      }
     }
   }
 
-  @TargetApi(29)
-  private void updateCurrentBitmap() {
+  @VisibleForTesting
+  /*package*/ Bitmap convertImageToBitmap(@NonNull Image image) {
+    Bitmap bitmap = null;
     if (android.os.Build.VERSION.SDK_INT >= 29) {
-      final HardwareBuffer buffer = currentImage.getHardwareBuffer();
-      currentBitmap = Bitmap.wrapHardwareBuffer(buffer, ColorSpace.get(ColorSpace.Named.SRGB));
+      final HardwareBuffer buffer = image.getHardwareBuffer();
+      bitmap = Bitmap.wrapHardwareBuffer(buffer, ColorSpace.get(ColorSpace.Named.SRGB));
```

Comment:
This should be super cheap right?

We're then paying the cost of synchronization/extra threading which will slow things down.